Add raja perf benchmarks

johnbowen42 opened the pull request.

johnbowen42 committed:
add benchmarks

johnbowen42 committed:
refactor driver and benchmark toml format

johnbowen42 committed:
Complete toml refactor and update script

johnbowen42 committed:
Change benchmarks toml more

johnbowen42 committed:
Bump submodule

johnbowen42 committed:
Fix driver and RAJAPerf cmake

johnbowen42 committed:
Make benchmark size smaller

johnbowen42 committed:
make toml problem size smaller to run faster on CI

johnbowen42 committed:
comment out everything but one benchmark

johnbowen42 committed:
Change to https for submodule

ggeorgakoudis commented on driver.py:
```diff
@@ -211,12 +220,10 @@ def build_and_run(self, reps, profiler=None):
 
-        exe = f"{self.benchmark}-{self.exemode}.x"
-        self.clean()
+        #self.clean()
```
IIRC for `hecbench` we have to clean first because otherwise different compilation modes (aot, proteus) pick up pre-built, incorrect object files. 

I think we should reinstate cleaning and add a `clean` key in the per benchmark or benchmark grouping if possible.

johnbowen42 committed:
Fix submodule url bug

ggeorgakoudis commented on driver.py:
```diff
@@ -142,10 +142,17 @@ def get_hash(x):
+        # the build command is meant to be a full bash command to build the benchmark, eg
+        # `cmake -DCMAKE_BUILD_TYPE=Debug --build` or `make benchmark`
+        # If none is provided, it will default to `make`
+        self.build_command = 'make' if build_command == None else build_command
```
Remove the default. User should define explicitly. We should use the TOML dictionary hierarchically, so user can provide a build command that applies to a group of programs.

ggeorgakoudis commented on driver.py:
```diff
@@ -461,19 +468,34 @@ def main():
                 args.compiler,
                 args.proteus_path,
-                env_configs,
+                env_configs
```
Use Black formatter or Ruff. It requires the comma here.

ggeorgakoudis commented on driver.py:
```diff
@@ -461,19 +468,34 @@ def main():
-
+    proteus_install = args.proteus_path
+    assert os.path.exists(proteus_install), f"Error: Proteus install path '{proteus_install}' does not exist!"
+    for env in env_configs:
+        env["PROTEUS_INSTALL_PATH"] = proteus_install
```
Where is that used?

johnbowen42 replied:
build instructions for RAJA perf

ggeorgakoudis commented on raja-perf-benchmarks.toml:
I assume programs are commented out because you are testing?

ggeorgakoudis commented on raja-perf-benchmarks.toml:
```diff
@@ -0,0 +1,1065 @@
+#[ATOMIC.nvidia.aot]
+#path = "benchmarks/RAJAPerf"
+#exe = "./build/bin/raja-perf.exe"
+#args = "--exclude-variants Base_Seq Lambda_Seq RAJA_Seq Base_CUDA RAJA_CUDA  "
```
Two questions:

1. Why do we do `--exclude-variants` instead of doing `--variant` to only run the desired one?
2. Why do you exclude `RAJA_CUDA`? That's the one we want to run without proteus enabled (same for HIP)

johnbowen42 replied:
1. Legacy, I can refactor but it's a little tedious
2. I am comparing `RAJA_CUDA` (which always uses proteus) to `Lambda_CUDA` in these benchmarks, as a hack to avoid modifying the benchmarks heavily to have an extra `RAJA_CUDA_JIT` variant.  `Lambda_CUDA` and `RAJA_CUDA` are roughly on parity (by design) for these benchmarks.  This can be fixed later and we can add a `RAJA_CUDA_JIT` variant to RAJA perf

ggeorgakoudis commented on raja-perf-benchmarks.toml:
```diff
@@ -0,0 +1,1065 @@
+[build]
+[build.nvidia]
```
We need build configurations with/without proteus

johnbowen42 replied:
see the above comment, I don't think we do.  Ultimately we will add `forall_jit` to `RAJA`, and a `RAJA_CUDA_JIT`/`RAJA_HIP_JIT` variant to RAJAPerf

johnbowen42 committed:
update toml

johnbowen42 committed:
Fix hecbench to work with new toml format and driver script.  Respond to feedback

johnbowen42 committed:
redo toml

johnbowen42 committed:
delete halo exchange because it requires mpi

johnbowen42 committed:
delete halo exchange because it requires mpi

johnbowen42 committed:
Update toml and script

johnbowen42 committed:
Update toml and script

johnbowen42 committed:
Fix unmangled name case in driver

tbennun commented on benchmarks.toml:
```diff
@@ -1,65 +1,129 @@
+[build]
+[build.nvidia]
```
(not for this PR) @johnbowen42 can we have build variants too? (e.g., build.nvidia.aot)

johnbowen42 committed:
Don't rebuild each bench

johnbowen42 committed:
change rep fact for all benchmarks

ggeorgakoudis committed:
Cleanup driver and toml spec

- Reduce replication in rajaperf toml
- Restore sizes and repfactor in rajaperf toml
- Use EAFP pattern in driver.py for dict accesses
- Memoize demangle for latency reduction

ggeorgakoudis committed:
Update plotting scripts

ggeorgakoudis committed:
Rename toml

ggeorgakoudis committed:
Fix hecbench toml and driver handling for clean

ggeorgakoudis committed:
Reduce batch size for compilation slowdown plot
